Mathlib/Algebra/Order/Monoid/LocallyFiniteOrder.lean

```diff
@@ -0,0 +1,177 @@
+/-
+Copyright (c) 2025 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.Algebra.Group.Subgroup.Ker
+import Mathlib.Algebra.Order.Group.Units
+import Mathlib.Algebra.Order.Hom.TypeTags
+import Mathlib.Algebra.Order.Ring.Int
+import Mathlib.Data.Nat.Cast.Order.Ring
+import Mathlib.Order.WellQuasiOrder
+import Mathlib.Tactic.Abel
+
+/-!
+
+# Locally Finite Linearly Ordered Abelian Groups
+
+## Main results
+- `LocallyFiniteOrder.orderAddMonoidEquiv`:
+  Any nontrivial linearly ordered abelian group that is locally finite is isomorphic to `ℤ`.
+- `LocallyFiniteOrder.orderMonoidEquiv`:
+  Any nontrivial linearly ordered abelian group that is locally finite is isomorphic to `ℤ`.
+- `LocallyFiniteOrder.orderMonoidEquivWithZero`:
+  Any nontrivial linearly ordered abelian group with zero that is locally finite
+  is isomorphic to `ℤᵐ⁰`.
+
+-/
+
+open Finset
+
+section Multiplicative
+
+variable {M : Type*} [CancelCommMonoid M] [LinearOrder M] [IsOrderedMonoid M] [LocallyFiniteOrder M]
+
+@[to_additive]
+lemma Finset.card_Ico_mul_right [ExistsMulOfLE M] (a b c : M) :
+    #(Ico (a * c) (b * c)) = #(Ico a b) := by
+  have : (Ico (a * c) (b * c)) = (Ico a b).map (mulRightEmbedding c) := by
+    ext x
+    simp only [mem_Ico, mem_map, mulRightEmbedding_apply]
+    constructor
+    · rintro ⟨h₁, h₂⟩
+      obtain ⟨d, rfl⟩ := exists_mul_of_le h₁
+      exact ⟨a * d, ⟨by simpa using h₁, by simpa [mul_right_comm a c d] using h₂⟩,
+        by simp_rw [mul_assoc, mul_comm]⟩
+    · aesop
+  simp [this]
+
+@[to_additive]
+lemma card_Ico_one_mul [ExistsMulOfLE M] (a b : M)
+    (ha : 1 ≤ a) (hb : 1 ≤ b) :
+    #(Ico 1 (a * b)) = #(Ico 1 a) + #(Ico 1 b) := by
+  have : Ico 1 b ∪ Ico (1 * b) (a * b) = Ico 1 (a * b) := by
+    simp [Ico_union_Ico, ha, hb, Right.one_le_mul ha hb]
+  rw [← this, Finset.card_union, Finset.card_Ico_mul_right]
+  simp [add_comm]
+
+end Multiplicative
+
+variable {M G : Type*} [AddCancelCommMonoid M] [LinearOrder M] [IsOrderedAddMonoid M]
+    [LocallyFiniteOrder M] [AddCommGroup G] [LinearOrder G]
+    [IsOrderedAddMonoid G] [LocallyFiniteOrder G]
+
+variable (G) in
+/-- The canonical embedding (as a monoid hom) from a linearly ordered cancellative monoid into `ℤ`.
```

Comment:
```suggestion
/-- The canonical embedding (as a monoid hom) from a linearly ordered cancellative additive monoid into `ℤ`.
```
(this might overflow the line)